### PR TITLE
Decorator based pydantic models generation with validation

### PR DESCRIPTION
**Pull request by gazorby**
Hi @peterschutt

Great work on this lib! It looks like solid fondation for starlite ecosystem ;)

I was experimenting with your improved dto implementation and did not found a way to add validation to generated pydantic models, so here it is.

The usage is as follow:

```python
class User(DeclarativeBase):
    id: Mapped[UUID] = mapped_column(
        default=uuid4, primary_key=True, info={"dto": Attrib(mark=dto.Mark.READ_ONLY)}
    )
    email: Mapped[str]
    password_hash: Mapped[str] = mapped_column(info={"dto": Attrib(mark=dto.Mark.SKIP)})

@dto.dto(User, Purpose.WRITE)
class UserCreate:
    @dto.validator("email")
    def val_email(cls, v):
        if "@" not in v:
            raise ValueError("Invalid email")
```

When using `mapper_bind=True` generated pydantic models can produce an SQLAlchemy instance from which they have been created:

```python
@dto.dto(User, Purpose.WRITE, mapper_bind=True)
class UserCreate:
    pass

user = UserCreate(email="foo@email.me")
user_model = user.mapper() # This is a User instance
session.add(user_model)
```

Tests are missing before this can be merged

### CODE REVIEW

**Comment by peterschutt** [comment]
Hey @gazorby, thanks for this PR, looks like some really interesting ideas. I'll review and get back to you asap!

File: src/starlite_saqlalchemy/dto.py [hunk 1]
```diff
@@ -72,9 +84,46 @@ class Attrib(NamedTuple):
...
+
 def _construct_field_info(elem: Column | RelationshipProperty, purpose: Purpose) -> FieldInfo:
     default = getattr(elem, "default", None)
-    if purpose is Purpose.READ or default is None:
+    nullable = getattr(elem, "nullable", False)
+    if purpose is Purpose.READ:
+        return FieldInfo(...)
+    if default is None:
+        if not nullable:
+            return FieldInfo(default=None)
```
Comment (peterschutt) [inline src/starlite_saqlalchemy/dto.py:126]: Is this correct? If default is `None` and column is not nullable, should it be ellipsis?

So this is distinguishing between an explicit `None` default and the attribute not existing, right? Couldn't hurt to use a sentinel value other than `None` in the `getattr()` call also.

File: src/starlite_saqlalchemy/dto.py [hunk 2]
```diff
@@ -152,7 +227,108 @@ class User(DeclarativeBase):
...
+                elem: Column | RelationshipProperty
+                try:
+                    elem = columns[key]
+                except KeyError:
+                    elem = relationships[key]
+                attrib = _get_dto_attrib(elem)
+                if _should_exclude_field(purpose, elem, exclude_, attrib):
+                    continue
+                type_ = type_hint
+                if key in relationships and key not in cls.__annotations__:
```
Comment (peterschutt) [inline src/starlite_saqlalchemy/dto.py:320]: What is the reasoning behind this bit?

**Review by peterschutt — COMMENTED** [review]
Thanks heaps for the PR and the work you've done here. I like the `dto.mark()` function as a shortcut for creating the info dict, and your logic to create the pydantic model in `dto.dto()` is definitely more thorough than I had it!

I'm especially a fan of the base class to produce the mapped models, I think that should be opt-out rather than opt in.

I'm not yet sold on the decorator pattern. I believe we could do something like this, but I haven't tried it out yet, figured I'd see what you thought first:

```python
def validate_email(): ...

class Model(orm.Base):
    email: Mapped[str] = mapped_column(info={"dto": dto.mark(validators=[validate_email])})

then, when we construct the DTO, we can add them to the model like:

```python
validators = {f"_{key}_{fn.__name__}": validator(key, allow_reuse=True)(fn) for fn in attrib.validators}
```

There may be a reason that won't work which I'm not seeing as I'm out of time to try it out for now, but interested to hear what you think.

The other thing that strikes me about the decorator, is that we aren't decorating subclasses of `BaseModel` - would that be an option do you think? And if so, would it allow the validators to be registered without us needing to wrap `validator` and keep the model validator cache? If so, it might be more flexible way to let users define a custom base class for the dto, that can include whatever validators etc that they want. E.g.,:

```python
class Model(orm.Base):
    key: Mapped[int]

@dto.decorate(Model)
class M(BaseModel):
    @validator("key", check_fields=False)
    def validate_thing(*args, **kwargs):
        return 1
```
Personally, I prefer to use the sqlalchemy validators as I don't only use the domain models in the context of the controllers, and keeps it a bit simpler in my opinion with only one way to do things. But, that's just me! Definitely happy to make it possible to do more sophisticated things with the DTO models too.

I've done a bit of work to the code as I've been going through this, so I'll PR back into your branch, LMK what you think.

File: src/starlite_saqlalchemy/dto.py [hunk 3]
```diff
@@ -72,9 +84,46 @@ class Attrib(NamedTuple):
...
+    __sqla_model__: type[DeclarativeBase]
+
+    class Config(BaseConfig):
+        """Config for MapperBind pydantic models."""
+
+    def __init_subclass__(cls, model: type[DeclarativeBase]) -> None:
+        cls.__sqla_model__ = model
+        return super().__init_subclass__()
+
+    def mapper(self) -> DeclarativeBase:
```
[the remark this answers is not in the document]
Comment (gazorby) [reply src/starlite_saqlalchemy/dto.py:105]: `to_mapper` would be a better name for sure

File: src/starlite_saqlalchemy/dto.py [hunk 4]
```diff
@@ -152,7 +227,108 @@ class User(DeclarativeBase):
...
+            for key, type_hint in type_hints.items():
+                elem: Column | RelationshipProperty
+                try:
+                    elem = columns[key]
+                except KeyError:
+                    elem = relationships[key]
+                attrib = _get_dto_attrib(elem)
+                if _should_exclude_field(purpose, elem, exclude_, attrib):
+                    continue
+                type_ = type_hint
```
[the remark this answers is not in the document]
Comment (gazorby) [reply src/starlite_saqlalchemy/dto.py:319]: Yes, I forgot that case. I removed it at firts as all my models use `MappedAsDataclass` mixin, so final type hints are unwrapped.

File: src/starlite_saqlalchemy/dto.py (hunk 1 above)
Comment (gazorby) [reply src/starlite_saqlalchemy/dto.py:126]: You're right, I found that bug too just after opening this PR, forgot to push.

File: src/starlite_saqlalchemy/dto.py (hunk 2 above)
Comment (gazorby) [reply src/starlite_saqlalchemy/dto.py:320]: I think it was for testing, but it should be recursive by default as `factory` is.

File: src/starlite_saqlalchemy/dto.py [hunk 5]
```diff
@@ -152,7 +227,108 @@ class User(DeclarativeBase):
...
+            fields: dict[str, tuple[Any, FieldInfo]] = {}
+            name = cls.__name__
+            namespace = {
+                **sys.modules[model.__module__].__dict__,
+                **sys.modules[cls.__module__].__dict__,
+            }
+            type_hints = {
+                **get_type_hints(model, localns=namespace),
+                **cls.__annotations__,
+            }
```
[the remark this answers is not in the document]
Comment (peterschutt) [reply src/starlite_saqlalchemy/dto.py:309]: This was also my thinking behind `Attrib.pydantic_field` - to allow to specify specific types for the DTO.

**Comment by gazorby** [comment]
Thanks for the detailed review @peterschutt!

I fully agree that not subclassing `BaseModel` is weird, ideally, model declaration should as close as possible to pydantic docs.
The main problem is that pydantic create fields in [metaclass](https://github.com/pydantic/pydantic/blob/1ac1199f08c42803dc7c5e6fae754afd53e3f1b4/pydantic/_internal/_model_construction.py#L138), and there is no clean way to dynamically add fields and validator afterwards. It's technically possible but a bit hacky and not documented (see https://github.com/pydantic/pydantic/issues/1937#issuecomment-695313040, https://github.com/pydantic/pydantic/issues/2076#issuecomment-719817049).

So just subclassing pydantic's `BaseModel` would still require to either dynamically add fields (hacky) or recreating a model with fields from sqlalchemy (same as what we are doing now).

So this leaves us with three options:

**Factory function that create model**
This is the current approach. What's bothers me is that validators need to be declared on sqlalchemy models, mixing up database layer with validation/business logic. Plus it add verbosity to sqlalchemy models, especially in combination with `MappedAsDataclass`.

Also, the current implementation does not allow overriding fields, although this can be easily added. But at this point I think that defining a class better meets the need.

**Using a decorator approach that also create a new model**
As you said, it's counter intuitive because we don't subclass anything. The main advantage over the factory is that it looks more like a pydantic model, with validator being declared directly on the class.

**Have our own `BaseModel`/`ModelMetaclass` infering fields from sqlalchemy models**
This would be the more transparant way from the user perspective I think, but would rely on undocumented pydantic internals. It could be something like this:

```python
from pydantic.main import BaseModel as BaseModel_
from pydantic.main import ModelMetaclass as ModelMetaClass_

def infer_sqlalchemy_fields(namespace: dict[str, Any]) -> dict[str, Any]:
    pass

class ModelMetaclass(ModelMetaClass_):
    def __new__(mcs, name, bases, namespace, **kwargs):
        namespace["__annotations__"] = infer_sqlalchemy_fields(namespace)
        return super().__new__(name, bases, namespace, **kwargs)


class BaseModel(BaseModel_, metaclass=ModelMetaclass):
    def to_mapped(self) -> DeclarativeBase:
        pass
```

**Comment by gazorby** [comment]
I will give a shot to the custom `BaseModel` to see how it looks like, should I do this in a new PR?

File: src/starlite_saqlalchemy/dto.py (hunk 2 above)
Comment (gazorby) [reply src/starlite_saqlalchemy/dto.py]: Oh, I remember now. The recursive implementation we have currently does not handle more complex case (that I'm facing) like `list[Model]`, `Optional[list[Model]]`, `list[A] | list[B]` etc. I think we should remove it untill we get something more robust.

**Comment by peterschutt** [comment]
I did some experimenting, I think the best approach is to have the decorator wrap subtypes of `BaseModel` and use the decorated class as base class for the produced DTO. If a field is already defined on the decorated class, `factory()` doesn't overwrite it, and support setting validators for both statically and dynamically added model fields (need `check_field=False` for validators of dynamically added fields which can easily be documented).

I PR'd to your branch again as I had the work there that I was experimenting with so figured I may as well send it through for you to comment on.

Cheers!

File: src/starlite_saqlalchemy/dto.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (peterschutt) [reply src/starlite_saqlalchemy/dto.py]: ...annnnd that thinking was wrong, as obv. can't set the type via `Field()` - added `Attrib.pydantic_type`.

**Comment by peterschutt** [comment]
Closing in favor of https://github.com/topsport-com-au/starlite-saqlalchemy/pull/139 - it has all this work in it, easier if we have a feature branch for this in repo so I can push directly to that.

**Pull request closed without merging**